README
mha4mysql-node - Master High Availability Manager and tools for MySQL (MHA) for automating master failover and fast master switch. This package contains utility scripts running on a MySQL server machine. See http://code.google.com/p/mysql-master-ha/wiki/TableOfContents?tm=6 for details.


debian/changelog
mha4mysql-node (0.58-0) squeeze; urgency=low

  * Several bugfixes
  * IPV6 support
  * super_read_only support
  * enforce_storage_engine support
  * Version 0.58
  * With the help of Marcelo Altmann <marcelo.altmann@percona.com>

 -- Kenny Gryp <kenny@goss.be>  Thu, 22 Mar 2018 00:00:00 +0000

mha4mysql-node (0.56-0) squeeze; urgency=low

  * (Note: All changelogs are written here: http://code.google.com/p/mysql-master-ha/wiki/ReleaseNotes )
  * Version 0.56

 -- Yoshinori Matsunobu <Yoshinori.Matsunobu@gmail.com>  Tue, 1 Apr 2014 00:00:00 +0000

mha4mysql-node (0.54-0) squeeze; urgency=low

  * Version 0.54

 -- Yoshinori Matsunobu <Yoshinori.Matsunobu@gmail.com>  Sat, 1 Dec 2012 00:00:00 +0000

mha4mysql-node (0.53) stable; urgency=low

  * Version 0.53

 -- Yoshinori Matsunobu <Yoshinori.Matsunobu@gmail.com>  Mon, 9 Jan 2012 00:00:00 +0000

mha4mysql-node (0.52) stable; urgency=low

  * Version 0.52 (0.51 was skipped)

 -- Yoshinori Matsunobu <Yoshinori.Matsunobu@gmail.com>  Fri, 16 Sep 2011 00:00:00 +0000

mha4mysql-node (0.50) stable; urgency=low

  * Version 0.50

 -- Yoshinori Matsunobu <Yoshinori.Matsunobu@gmail.com>  Sat, 23 Jul 2011 00:00:00 +0000


lib/MHA/NodeUtil.pm
#!/usr/bin/env perl

#  Copyright (C) 2011 DeNA Co.,Ltd.
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#   along with this program; if not, write to the Free Software
#  Foundation, Inc.,
#  51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

package MHA::NodeUtil;

use strict;
use warnings FATAL => 'all';

use Carp qw(croak);
use MHA::NodeConst;
use File::Path;
use Errno();
use Socket qw(NI_NUMERICHOST getaddrinfo getnameinfo);

sub create_dir_if($) {
  my $dir = shift;
  if ( !-d $dir ) {
    eval {
      print "Creating directory $dir.. ";
      mkpath($dir);
      print "done.\n";
    };
    if ($@) {
      my $e = $@;
      undef $@;
      if ( -d $dir ) {
        print "ok. already exists.\n";
      }
      else {
        croak "failed to create dir:$dir:$e";
      }
    }
  }
}

# Compare file checksum between local and remote host
sub compare_checksum {
  my $local_file  = shift;
  my $remote_path = shift;
  my $ssh_user    = shift;
  my $ssh_host    = shift;
  my $ssh_port    = shift;
  $ssh_port = 22 unless ($ssh_port);

  my $local_md5 = `md5sum $local_file`;
  return 1 if ($?);
  chomp($local_md5);
  $local_md5 = substr( $local_md5, 0, 32 );
  my $ssh_user_host = $ssh_user . '@' . $ssh_host;
  my $remote_md5 =
`ssh $MHA::NodeConst::SSH_OPT_ALIVE -p $ssh_port $ssh_user_host \"md5sum $remote_path\"`;
  return 1 if ($?);
  chomp($remote_md5);
  $remote_md5 = substr( $remote_md5, 0, 32 );
  return 2 if ( $local_md5 ne $remote_md5 );
  return 0;
}

sub file_copy {
  my $to_remote   = shift;
  my $local_file  = shift;
  my $remote_file = shift;
  my $ssh_user    = shift;
  my $ssh_host    = shift;
  my $log_output  = shift;
  my $ssh_port    = shift;
  $ssh_port = 22 unless ($ssh_port);
  my $dsn_host = $ssh_host =~ m{:} ? '[' . $ssh_host . ']' : $ssh_host;
  my $ssh_user_host = $ssh_user . '@' . $dsn_host;
  my ( $from, $to );
  if ($to_remote) {
    $from = $local_file;
    $to   = "$ssh_user_host:$remote_file";
  }
  else {
    $to   = $local_file;
    $from = "$ssh_user_host:$remote_file";
  }

  my $max_retries = 3;
  my $retry_count = 0;
  my $copy_fail   = 1;
  my $copy_command =
    "scp $MHA::NodeConst::SSH_OPT_ALIVE -P $ssh_port $from $to";
  if ($log_output) {
    $copy_command .= " >> $log_output 2>&1";
  }

  while ( $retry_count < $max_retries ) {
    if (
      system($copy_command)
      || compare_checksum(
        $local_file, $remote_file, $ssh_user, $ssh_host, $ssh_port
      )
      )
    {
      my $msg = "Failed copy or checksum. Retrying..";
      if ($log_output) {
        system("echo $msg >> $log_output 2>&1");
      }
      else {
        print "$msg\n";
      }
      $retry_count++;
    }
    else {
      $copy_fail = 0;
      last;
    }
  }
  return $copy_fail;
}

sub system_rc($) {
  my $rc   = shift;
  my $high = $rc >> 8;
  my $low  = $rc & 255;
  return ( $high, $low );
}

sub create_file_if {
  my $file = shift;
  if ( $file && ( !-f $file ) ) {
    open( my $out, ">", $file ) or croak "$!:$file";
    close($out);
  }
}

sub drop_file_if($) {
  my $file = shift;
  if ( $file && -f $file ) {
    unlink $file or croak "$!:$file";
  }
}

sub get_ip {
  my $host = shift;
  my ( $err, @bin_addr_host, $addr_host );
  if ( defined($host) ) {
    ( $err, @bin_addr_host ) = getaddrinfo($host);
    croak "Failed to get IP address on host $host: $err\n" if $err;

    # We take the first ip address that is returned by getaddrinfo
    ( $err, $addr_host ) = getnameinfo($bin_addr_host[0]->{addr}, NI_NUMERICHOST);
    croak "Failed to convert IP address for host $host: $err\n" if $err;

    # for IPv6 (and it works with IPv4 and hostnames as well):
    # - DBD-MySQL expects [::] format
    # - scp requires [::] format
    # - ssh requires :: format
    # - mysql tools require :: format
    # The code in MHA is expected to use [] when it is running scp
    # when it connects with DBD-MySQL

    return $addr_host;
  }
  return;
}

sub current_time() {
  my ( $sec, $min, $hour, $mday, $mon, $year ) = localtime();
  $mon  += 1;
  $year += 1900;
  my $val = sprintf( "%d-%02d-%02d %02d:%02d:%02d",
    $year, $mon, $mday, $hour, $min, $sec );
  return $val;
}

sub check_manager_version {
  my $manager_version = shift;
  if ( $manager_version < $MHA::NodeConst::MGR_MIN_VERSION ) {
    croak
"MHA Manager version is $manager_version, but must be $MHA::NodeConst::MGR_MIN_VERSION or higher.\n";
  }
}

sub parse_mysql_version($) {
  my $str = shift;
  my $result = sprintf( '%03d%03d%03d', $str =~ m/(\d+)/g );
  return $result;
}

sub parse_mysql_major_version($) {
  my $str = shift;
  $str =~ /(\d+)\.(\d+)/;
  my $strmajor = "$1.$2";
  my $result = sprintf( '%03d%03d', $strmajor =~ m/(\d+)/g );
  return $result;
}

sub mysql_version_ge {
  my ( $my_version, $target_version ) = @_;
  my $result =
    parse_mysql_version($my_version) ge parse_mysql_version($target_version)
    ? 1
    : 0;
  return $result;
}

my @shell_escape_chars = (
  '"', '!', '#', '&', ';', '`', '|',    '*',
  '?', '~', '<', '>', '^', '(', ')',    '[',
  ']', '{', '}', '$', ',', ' ', '\x0A', '\xFF'
);

sub unescape_for_shell {
  my $str = shift;
  if ( !index( $str, '\\\\' ) ) {
    return $str;
  }
  foreach my $c (@shell_escape_chars) {
    my $x       = quotemeta($c);
    my $pattern = "\\\\(" . $x . ")";
    $str =~ s/$pattern/$1/g;
  }
  return $str;
}

sub escape_for_shell {
  my $str = shift;
  my $ret = "";
  foreach my $c ( split //, $str ) {
    my $x      = $c;
    my $escape = 0;
    foreach my $e (@shell_escape_chars) {
      if ( $e eq $x ) {
        $escape = 1;
        last;
      }
    }
    if ( $x eq "'" ) {
      $x =~ s/'/'\\''/;
    }
    if ( $x eq "\\" ) {
      $x = "\\\\";
    }
    if ($escape) {
      $x = "\\" . $x;
    }
    $ret .= "$x";
  }
  $ret = "'" . $ret . "'";
  return $ret;
}

sub escape_for_mysql_command {
  my $str = shift;
  my $ret = "";
  foreach my $c ( split //, $str ) {
    my $x = $c;
    if ( $x eq "'" ) {
      $x =~ s/'/'\\''/;
    }
    $ret .= $x;
  }
  $ret = "'" . $ret . "'";
  return $ret;
}

sub client_cli_prefix {
  my ( $exe, $bindir, $libdir ) = @_;
  croak "unexpected client binary $exe\n" unless $exe =~ /^mysql(?:binlog)?$/;
  my %env = ( LD_LIBRARY_PATH => $libdir );
  my $cli = $exe;
  if ($bindir) {
    if ( ref $bindir eq "ARRAY" ) {
      $env{'PATH'} = $bindir;
    }
    elsif ( ref $bindir eq "" ) {
      $cli = escape_for_shell("$bindir/$exe");
    }
  }
  for my $k ( keys %env ) {
    if ( my $v = $env{$k} ) {
      my @dirs = ref $v eq "ARRAY" ? @{$v} : ( ref $v eq "" ? ($v) : () );
      @dirs = grep { $_ && !/:/ } @dirs;
      if (@dirs) {
        $cli = "$k="
          . join( ":", ( map { escape_for_shell($_) } @dirs ), "\$$k" )
          . " $cli";
      }
    }
  }

  # $cli .= " --no-defaults";
  return $cli;
}

1;


lib/MHA/SlaveUtil.pm
#!/usr/bin/env perl

#  Copyright (C) 2011 DeNA Co.,Ltd.
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#   along with this program; if not, write to the Free Software
#  Foundation, Inc.,
#  51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

package MHA::SlaveUtil;

use strict;
use warnings FATAL => 'all';

use Carp;
use English qw(-no_match_vars);
use File::Basename;
use MHA::NodeUtil;
use DBI;

use constant Get_Version_SQL => "SELECT VERSION() AS Value";
use constant Get_Datadir_SQL => "SELECT \@\@global.datadir AS Value";
use constant Get_Relay_Log_Info_Type_SQL =>
  "SELECT \@\@global.relay_log_info_repository AS Value";
use constant Get_Relay_Log_File_SQL =>
  "SELECT Relay_log_name FROM mysql.slave_relay_log_info";
use constant Get_Relay_Log_Info_SQL =>
  "SELECT \@\@global.relay_log_info_file AS Value";
use constant Is_Relay_Purge_SQL => "SELECT \@\@global.relay_log_purge As Value";
use constant Show_Log_Error_File_SQL => "SHOW VARIABLES LIKE 'log_error'";
use constant Show_Slave_Status_SQL   => "SHOW SLAVE STATUS";
use constant Disable_Relay_Purge_SQL => "SET GLOBAL relay_log_purge=0";
use constant Enable_Relay_Purge_SQL  => "SET GLOBAL relay_log_purge=1";
use constant Flush_Relay_Logs_SQL =>
  "FLUSH NO_WRITE_TO_BINLOG /*!50501 RELAY */ LOGS";
use constant Get_Failover_Lock_SQL =>
  "SELECT GET_LOCK('MHA_Master_High_Availability_Failover', ?) AS Value";
use constant Release_Failover_Lock_SQL =>
  "SELECT RELEASE_LOCK('MHA_Master_High_Availability_Failover') As Value";
use constant Get_Monitor_Lock_SQL =>
  "SELECT GET_LOCK('MHA_Master_High_Availability_Monitor', ?) AS Value";
use constant Release_Monitor_Lock_SQL =>
  "SELECT RELEASE_LOCK('MHA_Master_High_Availability_Monitor') As Value";

sub get_variable($$) {
  my $dbh   = shift;
  my $query = shift;
  my $sth   = $dbh->prepare($query);
  $sth->execute();
  my $href = $sth->fetchrow_hashref;
  return $href->{Value};
}

sub get_version($) {
  my $dbh = shift;
  return get_variable( $dbh, Get_Version_SQL );
}

sub get_log_error_file($) {
  my $dbh = shift;
  return get_variable( $dbh, Show_Log_Error_File_SQL );
}

sub get_relay_log_info_type {
  my $dbh           = shift;
  my $mysql_version = shift;
  my $type;
  $mysql_version = get_version($dbh) unless ($mysql_version);
  if ( !is_mariadb($mysql_version) && MHA::NodeUtil::mysql_version_ge( $mysql_version, "5.6.2" ) ) {
    $type = get_variable( $dbh, Get_Relay_Log_Info_Type_SQL );
  }
  unless ( defined($type) ) {
    $type = "FILE";
  }
  return $type;
}

sub get_relay_dir_file_from_table($) {
  my $dbh  = shift;
  my $sth  = $dbh->prepare(Get_Relay_Log_File_SQL);
  my $ret  = $sth->execute();
  my $href = $sth->fetchrow_hashref;
  if ( !defined($href) || !defined( $href->{Relay_log_name} ) ) {
    return;
  }
  my $current_relay_log_file = $href->{Relay_log_name};
  my $datadir = get_variable( $dbh, Get_Datadir_SQL );
  my $relay_dir;
  unless ( $current_relay_log_file =~ m/^\// ) {
    $current_relay_log_file =~ s/^\.\///;
    $current_relay_log_file = $datadir . "/" . $current_relay_log_file;
    $relay_dir              = $datadir;
    $relay_dir =~ s/\/$//;
  }
  else {
    $relay_dir = dirname($current_relay_log_file);
  }
  my $relay_log_basename = basename($current_relay_log_file);
  return ( $relay_dir, $relay_log_basename );
}

sub get_relay_log_info_path {
  my $dbh           = shift;
  my $mysql_version = shift;
  my $relay_log_info_path;
  my $filename;
  my $datadir = get_variable( $dbh, Get_Datadir_SQL );
  $mysql_version = get_version($dbh) unless ($mysql_version);
  if ( MHA::NodeUtil::mysql_version_ge( $mysql_version, "5.1.0" ) ) {
    $filename = get_variable( $dbh, Get_Relay_Log_Info_SQL );
  }

#relay_log_info_file was introduced in 5.1. In 5.0, it's fixed to "relay_log.info"
  if ( !defined($filename) ) {
    $filename = "relay-log.info";
  }
  unless ( $filename =~ m/^\// ) {
    $filename =~ s/^\.\///;
    $datadir =~ s/\/$//;
    $relay_log_info_path = $datadir . "/" . $filename;
  }
  else {
    $relay_log_info_path = $filename;
  }
  return $relay_log_info_path;
}

sub is_relay_log_purge($) {
  my $dbh = shift;
  return get_variable( $dbh, Is_Relay_Purge_SQL );
}

sub disable_relay_log_purge($) {
  my $dbh = shift;
  $dbh->do(Disable_Relay_Purge_SQL);
}

sub purge_relay_logs($) {
  my $dbh = shift;
  $dbh->do(Enable_Relay_Purge_SQL);
  $dbh->do(Flush_Relay_Logs_SQL);

  # To (almost) make sure that relay log is switched(purged) before setting
  # relay_log_purge=0;
  sleep 3;
  $dbh->do(Disable_Relay_Purge_SQL);
  return 0;
}

sub is_slave($) {
  my $dbh   = shift;
  my $query = Show_Slave_Status_SQL;
  my $sth   = $dbh->prepare($query);
  my $ret   = $sth->execute();
  if ( !defined($ret) || $ret != 1 ) {
    return 0;
  }
  return 1;
}

sub is_mariadb($) {
  my $version_str   = shift;
  if ($version_str =~ m/.*MariaDB.*/i)
  {
    return 1;
  }
  return 0;
}

sub get_advisory_lock_internal($$$) {
  my $dbh     = shift;
  my $timeout = shift;
  my $query   = shift;
  my $sth     = $dbh->prepare($query);
  my $ret     = $sth->execute($timeout);
  my $href    = $sth->fetchrow_hashref;
  if ( !defined($href) || !defined( $href->{Value} ) ) {
    return 2;
  }
  elsif ( $href->{Value} != 1 ) {
    return 1;
  }
  return 0;
}

sub release_advisory_lock_internal($$) {
  my $dbh   = shift;
  my $query = shift;
  my $sth   = $dbh->prepare($query);
  my $ret   = $sth->execute();
  my $href  = $sth->fetchrow_hashref;
  if ( !defined($href) || !defined( $href->{Value} ) || $href->{Value} != 1 ) {
    return 1;
  }
  return 0;
}

sub get_failover_advisory_lock {
  my $dbh     = shift;
  my $timeout = shift;
  return get_advisory_lock_internal( $dbh, $timeout, Get_Failover_Lock_SQL );
}

sub release_failover_advisory_lock($) {
  my $dbh = shift;
  return release_advisory_lock_internal( $dbh, Release_Failover_Lock_SQL );
}

sub get_monitor_advisory_lock {
  my $dbh     = shift;
  my $timeout = shift;
  return get_advisory_lock_internal( $dbh, $timeout, Get_Monitor_Lock_SQL );
}

sub release_monitor_advisory_lock($) {
  my $dbh = shift;
  return release_advisory_lock_internal( $dbh, Release_Monitor_Lock_SQL );
}

sub check_if_super_read_only {
  my $host = shift;
  my $ip = shift;
  my $port = shift;
  my $user = shift;
  my $pass = shift;
  # create databasehandle to check super_read_only setting
  my $dsn_host = $ip =~ m{:} ? '[' . $ip . ']' : $ip;
  my $dsn = "DBI:mysql:;host=$dsn_host;port=$port";
  my $dbh =
    DBI->connect( $dsn, $user, MHA::NodeUtil::unescape_for_shell($pass),
    { PrintError => 0, RaiseError => 1 } );
  croak "Failed to get DB Handle to check super_read_only on $host:$port!\n" unless ($dbh);

  my $sth = $dbh->prepare("SELECT \@\@global.super_read_only as Value");
  $sth->execute();
  my $href = $sth->fetchrow_hashref;
  if ( $href->{Value} == '1' ) {
    return (1, $dbh);
  }
  return (0, 0);
}

sub disable_super_read_only {
  my $dbh = shift;
  my $log = shift;
  my $message = "Disabling super_read_only, enabling read_only, so that the applying can be done on the slave\n";
  if ( $log ) {
    $log->info($message);
  } else {
    print $message;
  }
  $dbh->do("SET GLOBAL super_read_only=off;");
  $dbh->do("SET GLOBAL read_only=on;");
}

sub enable_super_read_only {
  my $dbh = shift;
  my $log = shift;
  my $message = "Enabling super_read_only again after applying\n";
  if ( $log ) {
    $log->info($message);
  } else {
    print $message;
  }
  $dbh->do("SET GLOBAL super_read_only=on;");
}

1;



lib/MHA/BinlogHeaderParser.pm
#!/usr/bin/env perl

#  Copyright (C) 2011 DeNA Co.,Ltd.
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#   along with this program; if not, write to the Free Software
#  Foundation, Inc.,
#  51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

package MHA::BinlogHeaderParser;

use strict;
use warnings FATAL => 'all';

use English qw(-no_match_vars);
use Carp qw(croak);
use File::Basename;
use File::Path;
use Errno();

sub new {
  my $class = shift;
  my $self  = {
    dir                        => undef,
    file                       => undef,
    self_server_id             => undef,
    target_rmlp                => undef,
    target_mlf                 => undef,
    master_version             => undef,
    current_mlf                => undef,
    starting_mlp               => undef,
    fp                         => undef,
    current_pos                => 4,
    prev_pos                   => 4,
    binlog_size                => undef,
    has_real_rotate_event      => 0,
    has_real_init_rotate_event => 0,
    has_checksum_algo          => 0,
    debug                      => 0,
    @_,
  };
  return bless $self, $class;
}

our $ROTATE_EVENT                 = 4;
our $FORMAT_DESCRIPTION_EVENT     = 15;
our $XID_EVENT                    = 16;
our $LOG_EVENT_MINIMAL_HEADER_LEN = 19;
our $MAX_ALLOWED_PACKET           = 1024 * 1024 * 1024;

sub open_binlog($) {
  my $self = shift;
  croak "binlog dir must be set.\n"  unless ( $self->{dir} );
  croak "binlog file must be set.\n" unless ( $self->{file} );
  open( $self->{fp}, "<", "$self->{dir}/$self->{file}" )
    or croak "$!:$self->{dir}/$self->{file}";
  binmode $self->{fp};
  $self->{binlog_size} = -s "$self->{dir}/$self->{file}";
}

sub close_binlog($) {
  my $self = shift;
  close( $self->{fp} );
}

sub unpack_header {
  my $self = shift;
  my $pos  = shift;
  $pos = $self->{current_pos} if ( !defined($pos) );
  my $after_pos = shift;
  $after_pos = $pos if ( !defined($after_pos) );
  my $fp        = $self->{fp};
  my $file_size = $self->{binlog_size};
  return unpack_header_util( $fp, $file_size, $pos, $after_pos );
}

sub unpack_header_util {
  my $fp        = shift;
  my $file_size = shift;
  my $pos       = shift;
  my $after_pos = shift;
  $after_pos = $pos if ( !defined($after_pos) );
  if ( $pos >= $file_size || $after_pos >= $file_size ) {
    warn "Invalid position is set. pos=$pos after_pos=$after_pos\n";
    return;
  }
  my $event_type;
  my $server_id;
  my $event_length;
  my $end_log_pos;
  seek( $fp, $pos + 4, 0 );
  read( $fp, $event_type, 1 );
  seek( $fp, $pos + 5, 0 );
  read( $fp, $server_id, 4 );
  seek( $fp, $pos + 9, 0 );
  read( $fp, $event_length, 4 );
  seek( $fp, $pos + 13, 0 );
  read( $fp, $end_log_pos, 4 );
  seek( $fp, $after_pos, 0 );
  $event_type   = unpack( 'C', $event_type );
  $server_id    = unpack( 'V', $server_id );
  $event_length = unpack( 'V', $event_length );
  $end_log_pos  = unpack( 'V', $end_log_pos );

  if ( $event_length > $MAX_ALLOWED_PACKET ) {
    croak
"Event too large: pos: $pos, event_length: $event_length, event_type: $event_type\n";
  }

  if ( $event_length < $LOG_EVENT_MINIMAL_HEADER_LEN ) {
    croak
"Event too small: pos: $pos, event_length: $event_length, event_type: $event_type\n";
  }
  return ( $event_type, $server_id, $event_length, $end_log_pos );
}

sub get_server_version_from_fde($$) {
  my $self      = shift;
  my $pos       = shift;
  my $is_master = shift;
  my $server_version;
  my $fp = $self->{fp};
  my ( $event_type, $server_id, $event_length, $c ) =
    $self->unpack_header($pos);
  if ( $event_type == $FORMAT_DESCRIPTION_EVENT ) {
    seek( $fp, $pos + 21, 0 );
    read( $fp, $server_version, 50 );
    $server_version =~ s/\0//g;
    print " Master Version is $server_version\n" if ($is_master);
    if ( MHA::NodeUtil::mysql_version_ge( $server_version, "5.6.0" ) ) {
      seek( $fp, $pos + $event_length - 5, 0 );
      read( $fp, $self->{checksum_algo}, 1 );
      if ( $self->{checksum_algo} ne "\0" ) {
        print " Binlog Checksum enabled\n";
        $self->{checksum_algo} = 1;
      }
      else {
        $self->{checksum_algo} = 0;
      }
    }
    if ($is_master) {
      $self->{master_version} = $server_version;
    }
  }
  else {
    croak
      "This is not format description event! ev type=$event_type, pos=$pos.\n";
  }
  seek( $fp, $pos, 0 );
  return $server_version;
}

sub parse_master_rotate_event($$) {
  my $self = shift;
  my $pos  = shift;
  my $fp   = $self->{fp};
  my ( $event_type, $server_id, $event_length, $end_log_pos ) =
    $self->unpack_header($pos);

  if ( $event_type == $ROTATE_EVENT ) {

    # real rotate event
    if ( $end_log_pos == 0 || $end_log_pos == $self->{current_pos} ) {
      my $offset     = 19;
      my $header_len = 8;
      my $rotate_pos;
      seek( $fp, $pos + $offset, 0 );
      read( $fp, $rotate_pos, $header_len );
      $end_log_pos = unpack( 'V', $rotate_pos );
      seek( $fp, $pos + $offset + $header_len, 0 );
      my $read_length = $event_length - $offset - $header_len;

      if ( $self->{checksum_algo} ) {
        $read_length -= 4;
      }
      read( $fp, $self->{current_mlf}, $read_length );
      $self->{has_real_rotate_event} = 1;
      $self->{has_real_init_rotate_event} = 1 if ( $end_log_pos <= 4 );
      print
"parse_master_rotate_event(real rotate event): file=$self->{file} event_type=$event_type server_id=$server_id length=$event_length elp=$end_log_pos\n"
        if ( $self->{debug} );
    }
    else {
      print
"parse_master_rotate_event(non-real rotate event): file=$self->{file} event_type=$event_type server_id=$server_id length=$event_length elp=$end_log_pos\n"
        if ( $self->{debug} );
    }
  }
  else {
    croak "This is not rotate event! ev type=$event_type, pos=$pos.\n";
  }
  seek( $fp, $pos, 0 );
  $self->{starting_mlp} = $end_log_pos unless ( $self->{starting_mlp} );
  return $end_log_pos;
}

sub parse_forward_header($) {
  my $self = shift;
  my $fp   = $self->{fp};
  my $pos  = $self->{current_pos};
  return if ( $self->{current_pos} >= $self->{binlog_size} );

  my ( $event_type, $server_id, $event_length, $end_log_pos ) =
    $self->unpack_header($pos);
  $self->{prev_pos}    = $self->{current_pos};
  $self->{current_pos} = $self->{current_pos} + $event_length;
  return ( $event_type, $server_id, $event_length, $end_log_pos );
}

sub parse_init_headers {
  my $self    = shift;
  my $do_dump = shift;
  my $outfile = shift;

  while ( my ( $event_type, $server_id, $event_length, $end_log_pos ) =
    $self->parse_forward_header() )
  {
    print
"parse_init_headers: file=$self->{file} event_type=$event_type server_id=$server_id length=$event_length nextmpos=$end_log_pos prevrelay=$self->{prev_pos} cur(post)relay=$self->{current_pos}\n"
      if ( $self->{debug} );
    if ( $event_type == $FORMAT_DESCRIPTION_EVENT ) {
      if ( $self->{self_server_id} && $server_id != $self->{self_server_id} ) {
        $self->{starting_mlp} = $end_log_pos
          if ( !$self->{starting_mlp} && $end_log_pos > 0 );
        if ( !$self->{master_version} ) {
          $self->get_server_version_from_fde( $self->{prev_pos}, 1 );
        }
      }
      else {
        $self->get_server_version_from_fde( $self->{prev_pos}, 0 );
      }
    }
    elsif ( $event_type == $ROTATE_EVENT ) {
      if ( $self->{self_server_id} && $server_id != $self->{self_server_id} ) {
        $end_log_pos = $self->parse_master_rotate_event( $self->{prev_pos} );
      }
      if ($do_dump) {
        print
"dumping rotate event from $self->{prev_pos} to $self->{current_pos}.\n"
          if ( $self->{debug} );
        my $buf;
        open( my $out, ">>", $outfile ) or croak "$!:$outfile";
        binmode $out;
        seek( $self->{fp}, $self->{prev_pos}, 0 );
        read( $self->{fp}, $buf, $self->{current_pos} - $self->{prev_pos} );
        print $out $buf;
        close($out);
      }
    }
    else {
      return $self->{prev_pos};
    }
  }
  return $self->{current_pos};
}

sub get_starting_mlp($) {
  my $self = shift;
  if ( $self->{starting_mlp} ) {
    print "starting_mlp: $self->{starting_mlp}\n" if ( $self->{debug} );
    return $self->{starting_mlp};
  }
  while ( my ( $event_type, $server_id, $event_length, $end_log_pos ) =
    $self->parse_forward_header() )
  {
    print
"get_starting_mlp: file=$self->{file} event_type=$event_type server_id=$server_id length=$event_length next=$end_log_pos\n"
      if ( $self->{debug} );
    next if ( $server_id == $self->{self_server_id} );
    next if ( $event_type == $FORMAT_DESCRIPTION_EVENT );
    next if ( $event_type == $ROTATE_EVENT );
    next if ( $end_log_pos <= 0 );
    $self->{starting_mlp} = $end_log_pos;
    return $self->{starting_mlp};
  }
  return;
}

sub get_starting_mlp_xid($) {
  my $self = shift;
  while ( my ( $event_type, $server_id, $event_length, $end_log_pos ) =
    $self->parse_forward_header() )
  {
    print
"get_starting_mlp_xid: file=$self->{file} event_type=$event_type server_id=$server_id length=$event_length next=$end_log_pos\n"
      if ( $self->{debug} );
    next if ( $server_id == $self->{self_server_id} );
    next if ( $event_type == $FORMAT_DESCRIPTION_EVENT );
    next if ( $event_type == $ROTATE_EVENT );
    next if ( $end_log_pos <= 0 );
    if ( $event_type == $XID_EVENT ) {
      $self->{starting_mlp} = $end_log_pos;
      return $self->{starting_mlp};
    }
    else {
      if ( $self->{starting_mlp} && $self->{starting_mlp} <= $end_log_pos ) {
        return $self->{starting_mlp};
      }
      else {
        return;
      }
    }
  }
  return;
}

sub find_target_relay_pos($) {
  my $self  = shift;
  my $found = 0;
  while ( my ( $event_type, $server_id, $event_length, $end_log_pos ) =
    $self->parse_forward_header() )
  {
    print
"find_target_relay_pos: file=$self->{file} event_type=$event_type server_id=$server_id length=$event_length next=$end_log_pos\n"
      if ( $self->{debug} );
    next if ( $server_id == $self->{self_server_id} );
    next if ( $event_type == $FORMAT_DESCRIPTION_EVENT );
    if ( $event_type == $ROTATE_EVENT ) {
      $end_log_pos = $self->parse_master_rotate_event( $self->{prev_pos} );
    }
    if ( $end_log_pos == $self->{target_rmlp} ) {
      $found = 1;
      return ( $found, $self->{current_pos} );
    }
  }
  $found = 0;
  return $found;
}

sub find_target_relay_pos_xid($$$$$) {
  my $self               = shift;
  my $last_absolute_mpos = shift;
  my $prev_mpos          = shift;
  my $in_relative        = shift;
  my $real_mpos          = shift;

  my $found = 0;
  my $start_rlp;

  while ( my ( $event_type, $server_id, $event_length, $end_log_pos ) =
    $self->parse_forward_header() )
  {
    print
"find_target_relay_pos_xid: file=$self->{file} event_type=$event_type server_id=$server_id length=$event_length next=$end_log_pos\n"
      if ( $self->{debug} );
    next if ( $event_type == $FORMAT_DESCRIPTION_EVENT );
    next if ( $server_id == $self->{self_server_id} );
    if ( $event_type == $ROTATE_EVENT ) {
      $end_log_pos = $self->parse_master_rotate_event( $self->{prev_pos} );
    }
    if ( $end_log_pos == $event_length
      || ( $end_log_pos < $prev_mpos && !$in_relative ) )
    {

      # Entering relative mode
      $last_absolute_mpos = $prev_mpos;
      $in_relative        = 1;
      $real_mpos          = $last_absolute_mpos + $end_log_pos;
    }
    elsif ( $end_log_pos < $prev_mpos && $in_relative ) {

      # void event
    }
    elsif ( $in_relative && $event_type != $XID_EVENT ) {

      # Continuing relative mode
      $real_mpos = $last_absolute_mpos + $end_log_pos;
    }
    elsif ( $event_type == $XID_EVENT ) {

      # Exiting relative mode
      $real_mpos          = $end_log_pos;
      $last_absolute_mpos = $end_log_pos;
      $in_relative        = 0;
    }
    else {

      # Not in relative mode
      $real_mpos          = $end_log_pos;
      $last_absolute_mpos = $end_log_pos;
    }
    $prev_mpos = $end_log_pos;
    if ( $real_mpos == $self->{target_rmlp} ) {
      $found     = 1;
      $start_rlp = $self->{current_pos};
      last;
    }
  }
  return ( $found, $start_rlp, $last_absolute_mpos, $prev_mpos, $in_relative,
    $real_mpos );
}

1;



lib/MHA/NodeConst.pm
#!/usr/bin/env perl

#  Copyright (C) 2011 DeNA Co.,Ltd.
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#   along with this program; if not, write to the Free Software
#  Foundation, Inc.,
#  51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

use strict;
use warnings FATAL => 'all';

package MHA::NodeConst;

our $VERSION         = '0.58';
our $MGR_MIN_VERSION = '0.54';
our $SSH_OPT_ALIVE =
"-o ServerAliveInterval=60 -o ServerAliveCountMax=20 -o StrictHostKeyChecking=no -o ConnectionAttempts=5 -o PasswordAuthentication=no -o BatchMode=yes";

# apply_diff_relay_logs status codes
our $Target_Has_Received_All_Relay_Logs = 2;
our $Relay_Log_Not_Found                = 10;
our $Relay_Pos_Not_Found                = 12;
our $Applying_SQL_File_Failed           = 22;

1;
